### ISSUE TITLE
plot_scene_types.r  |  Error: Insufficient values in manual scale. 15 needed but only 2 provided.

### ISSUE THREAD

**Issue by aj1s** [TNC-NMFO/NWLAND#9]
Incidentally, in the text for test runs in the CALAND documentation (https://github.com/aldivi/caland#instructions-for-your-first-test-runs-with-caland), a dbl quote is needed before 'Water':

plot_scen_types(varname = "All_orgC_stock_diff", ylabel = "Change from Baseline (MMTC)", data_dir = "./outputs", file_tag = "", reg = c("All_region"), lt = c(Water", "Ice", "Barren", "Sparse", "Desert", "Shrubland", "Grassland", "Savanna", "Woodland", "Forest", "Meadow", "Coastal_marsh", "Fresh_marsh", "Cultivated", "Developed_all"), own = c("All_own"), figdir = "figures")

**Comment by aj1s** [comment]
@aj1s

**Comment by aj1s** [comment]
test @aj1s

**Comment by aj1s** [comment]
@msimmond

hi Maegen –

This error occurred in the course of following the CALAND test run instructions (https://github.com/aldivi/caland#instructions-for-your-first-test-runs-with-caland) when using plot_caland.r and plot_scene_types.r as modified by Issues #7 and #8 respectively. After running CALAND() on each of the two example scenarios and the modified plot_caland() for the plotted comparison of the two, the error is thrown when attempting to run the modified version of plot_scene_types().

Would it help for me to provide you with my R workspace ? 

My system / envi info:
Dell Precision 7750, Intel(R) Core(TM) i7-10875H CPU @ 2.30GHz   2.30 GHz
Running Windows 10, 64-bit
R version (64-bit):  4.0.3
RStudio version 1.3.1093

**Comment by msimmond** [comment]
It would help to see your full workflow with the arguments used, from write_caland_inputs() to the point where the error occurred, in this case, to plot_scen_types()

**Comment by aldivi** [comment]
This sounds like an issue related to assigning levels for plotting. Check:
scale_shape_manual(values=1:nlevels(plot_df$Land_Type))

**Comment by aj1s** [comment]
The workflow leading to this error begins with the test input files distributed with this zip archive: https://zenodo.org/record/3256727/files/aldivi/caland-v3.0.0.zip and thus skips `write_caland_inputs()`.  I first load the three functions, `CALAND()`, `plot_caland()`, and `plot_scen_types()`, then run in sequence the test sets of arguments as described here: https://github.com/aldivi/caland#instructions-for-your-first-test-runs-with-caland. (I do so verbatim, except for the addition of the double quote before ‘water’ in the arguments for running `plot_scen_types()`.)
  
`CALAND(scen_file="NWL_Historical_v6_default_RCP85.xls", c_file_arg = "carbon_input_nwl.xls", indir = "", outdir = "", start_year = 2010, end_year = 2101, value_col_dens = 7, ADD_dens = TRUE, value_col_accum = 7, ADD_accum = TRUE, value_col_soilcon=8, ADD_soilcon = TRUE, NR_Dist = 120, WRITE_OUT_FILE = TRUE, blackC = FALSE)`

`CALAND(scen_file="NWL_Alt_A_v6_default_RCP85.xls", c_file_arg = "carbon_input_nwl.xls", indir = "", outdir = "", start_year = 2010, end_year = 2101, value_col_dens = 7, ADD_dens = TRUE, value_col_accum = 7, ADD_accum = TRUE, value_col_soilcon=8, ADD_soilcon = TRUE, NR_Dist = -1, WRITE_OUT_FILE = TRUE, blackC = FALSE)`

`plot_caland(scen_fnames = c("NWL_Historical_v6_default_RCP85_output_mean_BC1_NR120.xls", "NWL_Alt_A_v6_default_RCP85_output_mean_BC1.xls"), scen_snames = c("Baseline","A"), reg = c("All_region"), lt = c("All_land", "Water", "Ice", "Barren", "Sparse", "Desert", "Shrubland", "Grassland", "Savanna", "Woodland", "Forest", "Meadow", "Coastal_marsh", "Fresh_marsh", "Cultivated", "Developed_all", "Seagrass"), own=c("All_own"))`

`plot_scen_types(varname = "All_orgC_stock_diff", ylabel = "Change from Baseline (MMTC)", data_dir = "./outputs", file_tag = "", reg = c("All_region"), lt = c("Water", "Ice", "Barren", "Sparse", "Desert", "Shrubland", "Grassland", "Savanna", "Woodland", "Forest", "Meadow", "Coastal_marsh", "Fresh_marsh", "Cultivated", "Developed_all"), own = c("All_own"), figdir = "figures")`

**Note**: in the last argument set, the double-quote preceding "Water" is missing from the corresponding text in the CALANDv3_User_Guide.pdf.

It is on running `plot_scen_types()` that I receive the error, and  `<rlang::last_trace()>` returns the following:

```
  1. \-global::plot_scen_types(...)
  2.   +-base::do.call(...)
  3.   \-(function (filename, plot = last_plot(), device = NULL, path = NULL, ...
  4.     +-grid::grid.draw(plot)
  5.     \-ggplot2:::grid.draw.ggplot(plot)
  6.       +-base::print(x)
  7.       \-ggplot2:::print.ggplot(x)
  8.         +-ggplot2::ggplot_build(x)
  9.         \-ggplot2:::ggplot_build.ggplot(x)
 10.           \-base::lapply(data, scales_map_df, scales = npscales)
 11.             \-ggplot2:::FUN(X[[i]], ...)
 12.               +-base::unlist(...)
 13.               \-base::lapply(scales$scales, function(scale) scale$map_df(df = df))
 14.                 \-ggplot2:::FUN(X[[i]], ...)
 15.                   \-scale$map_df(df = df)
 16.                     \-ggplot2:::f(..., self = self)
 17.                       \-base::lapply(aesthetics, function(j) self$map(df[[j]]))
 18.                         \-ggplot2:::FUN(X[[i]], ...)
 19.                           \-self$map(df[[j]])
 20.                             \-ggplot2:::f(..., self = self)
 21.                               \-self$palette(n)
 22.                                 \-ggplot2:::f(...)
```

I created this issue after having looked – to no avail of my own – at this recurring argument:

`scale_fill_manual(values = ghg_colors, breaks = breaks)`

with reference to the top-voted solution here: https://stackoverflow.com/questions/18999284/ggplot2-error-insufficient-values-in-manual-scale

However, I had overlooked the line Alan highlighted: 

`scale_shape_manual(values=1:nlevels(plot_df$Land_Type))`


Would it help if I share my R project through RStudio Cloud ?

**Comment by aj1s** [comment]
@aldivi 

Hi Alan, I’m still not able to determine why scale_shape_manual() isn’t able to map values from all 15 csvs referenced by this test argument set for plot_scene_types(), all of which I’ve confirmed are populated and available in the outputs directory (e.g., All_region_Water_All_own_All_orgC_stock_diff_output.csv    
All_region_Ice_All_own_All_orgC_stock_output.csv, ...) from the preceding plot_caland() arguments.		

I should note as well this error is also thrown when the test runs use the caland-v3.0.0 code. 

When you have a chance, could you have a closer look, yourself ? Thank you

**Comment by aldivi** [comment]
@aj1s I can't reproduce this error. I am using ggplot 3.3.2. What version do you have?
Can you send me the required diagnostic files from plot_caland() so I can try them? All I really need are the 'All_orgC_stock_diff' csv files for each of the land types in your All_region diagnostics folder.
If it is easier you can zip up the entire All_region folder with all of your plot_caland() diagnostic outputs and send that to me.

**Comment by aldivi** [comment]
Also, where is the modified version of plot_scen_types that you are using?

**Comment by aj1s** [comment]
@aldivi  –  Thanks for having a(nother) look at this. 

I’m using ggplot 3.3.3.  

The modified version of ‘plot_scen_types.r’ is:
https://github.com/TNC-NMFO/NWLAND/blob/nwland_dev/plot_scen_types.r
(Note: The only change I made was increasing the memory available to Java.)

The ‘All_orgC_Stock_diff’ csvs and two main scenario output files from the test run (argument sets defined above in this Issue) are in this zip I’ve sent by e-mail: 

Issue9_Diag_files_from test_run_of plot_caland_20210513.zip 

(The zip archive is 108.8mb, and I haven’t yet installed git LFS.)

**Comment by aj1s** [comment]
@aldivi –   After finding (on line 148) that setting scale_shape_manual() outside of aes() didn’t resolve the issue, I then tried using ggplot2 v 3.3.2: 

```
library(remotes)
remotes::install_version("ggplot2", version = "3.3.2")

```
...but received the original error. 

Next I tried a run with line 148 simply commented out (as I can still read the plots with the geoms distinguished only by color). This returned:

```
Warning messages:
1: The shape palette can deal with a maximum of 6 discrete values because more than 6 becomes difficult to discriminate; you have 15. Consider specifying shapes manually if you must have them. 
2: Removed 378 rows containing missing values (geom_point).
```

So, I changed line 148 to assign shapes based on explicit known count of land_types rather than by that of factor levels:

	+ scale_shape_manual(values = c(1:18))

and this returned the desired plot. Given that I don’t think the count of land types is likely to change (or, not often) between our scenarios run over all of CO and NM, and given the lower priority of this Issue, I’m content closing it with this fix.
